platform_disk_api/admission_controller/api.py

```diff
@@ -1,43 +1,316 @@
+            self._kube_client.generate_namespace_url(namespace)
         )
+        try:
+            namespace_labels = namespace_obj["metadata"]["labels"]
+            org = namespace_labels[APOLO_ORG_LABEL]
+            project = namespace_labels[APOLO_PROJECT_LABEL]
+        except KeyError:
+            return admission_review.decline(
+                status_code=HTTPBadRequest.status_code,
+                message="Namespace lacks required org / project labels",
+            )
 
+        now = datetime_dump(utc_now())
 
-async def create_app(config: Config) -> aiohttp.web.Application:
-    app = aiohttp.web.Application()
+        if "annotations" not in pvc_metadata:
+            LOGGER.info("PVC doesn't define any annotation. Going to create")
+            admission_review.add_patch(PATH_ANNOTATIONS, value={})
 
-    AdmissionControllerHandler(app=app).register()
+        if "labels" not in pvc_metadata:
+            LOGGER.info("PVC doesn't define any label. Going to create")
+            admission_review.add_patch(PATH_LABELS, value={})
+
+        # populate necessary disk annotations
+        for annotation_key, annotation_value in (
+            (APOLO_DISK_API_CREATED_AT_ANNOTATION, now),
+            (DISK_API_CREATED_AT_ANNOTATION, now),
+        ):
+            self._add_key_value_if_not_exist(
+                admission_review=admission_review,
+                collection=pvc_annotations,
+                collection_path=PATH_ANNOTATIONS,
+                key=annotation_key,
+                value=annotation_value,
+            )
+
+        # populate necessary disk labels
+        for label_key, label_value in (
+            (DISK_API_MARK_LABEL, "true"),
+            (APOLO_DISK_API_MARK_LABEL, "true"),
+            (DISK_API_ORG_LABEL, org),
+            (APOLO_ORG_LABEL, org),
+            (DISK_API_PROJECT_LABEL, project),
+            (APOLO_PROJECT_LABEL, project),
+            (APOLO_USER_LABEL, project),
+            (USER_LABEL, project),
+        ):
+            self._add_key_value_if_not_exist(
+                admission_review=admission_review,
+                collection=pvc_labels,
+                collection_path=PATH_LABELS,
+                key=label_key,
+                value=label_value,
+            )
 
+        LOGGER.info(f"Will submit patch operations: {admission_review.patch}")
+
+        # if a disk was created via an API with the name provided,
+        # it'll be stored in those annotations.
+        disk_name = pvc_annotations.get(
+            APOLO_DISK_API_NAME_ANNOTATION
+        ) or pvc_annotations.get(DISK_API_NAME_ANNOTATION)
+
+        # create a disk naming
+        disk_name = Service.get_disk_naming_name(
+            disk_name or pvc_name,
+            org_name=org,
+            project_name=project,
+        )
+        disk_naming = DiskNaming(namespace, name=disk_name, disk_id=pvc_name)
+
+        try:
+            await self._kube_client.create_disk_naming(disk_naming)
```

Comment:
Can we not always fail here? For instance this is a retry request for the disk and we already created disk naming in the previous request

Reply:
@zubenkoivan changed, it should work now in a case of a reinvocation.